### PR TITLE
feat: support Expand Equal ONNX op

### PR DESCRIPTION
**Pull request by jfrery**
ref https://github.com/zama-ai/concrete-ml-internal/issues/4158

I had to implement Equal as well because the `torch.norm(...)` seems to create an Equal node. It's only there to compute constants so it does not appear in the final FHE graph.

### CODE REVIEW

File: src/concrete/ml/onnx/ops_impl.py [hunk 1]
```diff
@@ -2043,3 +2062,24 @@ def numpy_gather(
     slices = tuple(slice(None) if i != axis else indices_list for i in range(x.ndim))
 
     return (x[slices],)
+
+
+@onnx_func_raw_args("shape")
+def numpy_expand(x: numpy.ndarray, shape=None) -> Tuple[numpy.ndarray]:
```
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/ops_impl.py:2068]: shape parameter is missing a typing (I guess `Optional[Tuple[int]]`) !

File: src/concrete/ml/onnx/ops_impl.py [hunk 2]
```diff
@@ -2043,3 +2062,24 @@ def numpy_gather(
...
+    """Apply the expand operator in numpy according to ONNX spec.
+
+    See https://github.com/onnx/onnx/blob/main/docs/Operators.md#expand
+
+    Args:
+        x (numpy.ndarray): Input tensor.
+        shape: Tuple of the new shape.
```
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/ops_impl.py:2075]: same, missing typing !

File: src/concrete/ml/onnx/ops_impl.py [hunk 3]
```diff
@@ -2043,3 +2062,24 @@ def numpy_gather(
...
+
+    Returns:
+        Tuple[numpy.ndarray]: Output tensor.
+    """
+    target_shape = numpy.array(shape, dtype=int)
+    padding = len(target_shape) - len(x.shape)
```
Comment (RomanBredehoft) [inline src/concrete/ml/onnx/ops_impl.py:2081]: `padding` is a confusing name I would say, maybe `shape_difference` ? not very important anyway

File: src/concrete/ml/quantization/quantized_ops.py [hunk 4]
```diff
@@ -2430,3 +2430,87 @@ def can_fuse(self) -> bool:
...
+            op_instance_name,
+            int_input_names,
+            constant_inputs,
+            input_quant_opts,
+            **attrs,
+        )
+
+        # We do not support testing a == b where a,b are encrypted
+        # only comparing to a constant is supported
+        assert_true(constant_inputs is not None and len(constant_inputs) >= 1)
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_ops.py:2516]: I'm wondering, shouldn't it be `len(constant_inputs) == 1` (or `len(constant_inputs) == 2`) ? why are we making this check "so loose" ?

File: tests/quantization/test_quantized_ops.py [hunk 5]
```diff
@@ -1146,6 +1148,60 @@ def test_quantized_reshape(shape):
     )
 
 
+@pytest.mark.parametrize(
+    "original_shape, expand_shape",
+    [
```
Comment (RomanBredehoft) [inline tests/quantization/test_quantized_ops.py:1153]: great thorough tests ! But maybe there was a way to automate this ? How can we be sure that all interesting cases are covered ?

File: tests/quantization/test_quantized_ops.py [hunk 6]
```diff
@@ -1193,81 +1249,66 @@ def test_quantized_prelu(n_bits, input_range, input_shape, slope, is_signed, che
     )
 
 
-@pytest.mark.parametrize(
-    "params",
-    [
+@pytest.fixture(scope="module")
+def random_test_data():
+    """Generate data for comparisons operators."""
+    numpy.random.seed(42)  # Fixing the seed for reproducibility
```
Comment (RomanBredehoft) [inline tests/quantization/test_quantized_ops.py:1255]: shouldn't we make it random like other tests : `random_state = numpy.random.randint(0, 2**15)` ? If we fix seeds like this, we'll never encounter some possible corner cases that would need to be fixed !

File: tests/torch/test_compile_torch.py [hunk 7]
```diff
@@ -1144,7 +1147,12 @@ def test_shape_operations_net(
     check_float_array_equal,
 ):
     """Test a pattern of reshaping, concatenation, chunk extraction."""
-    net = ShapeOperationsNet(is_qat)
+    net = model(is_qat)
```
Comment (RomanBredehoft) [inline tests/torch/test_compile_torch.py:1150]: small detail but in most of our tests we use `model_class` for the parameter and `model` for the actual instance (instead of `net` here), would be great if you could update this test with that !

File: tests/torch/test_compile_torch.py [hunk 8]
```diff
@@ -1144,7 +1147,12 @@ def test_shape_operations_net(
...
+    if is_qat and not net.is_qat_compatible:
+        pytest.skip(f"Model {model.__name__} is not compatible with is_qat=True")
+
+    # Shape transformation do not support >1 example in the inputset
+    # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/3871
     inputset = numpy.random.uniform(size=(1, n_channels, 2, 2))
 
     if is_qat:
```
Comment (RomanBredehoft) [inline tests/torch/test_compile_torch.py:1158]: just realized by reading what's following that in case of `is_qat=False`, we never check any r2 score or whatever. Basically, this PR is green only because the forward/compile didn't fail ! There is only this comment bu no FIXME attached to it : is this still expected ? : 
```
    # In QAT quantization options are consistent across all the layers
    # which allows for the elimination of TLUs
    # In PTQ there are TLUs in the graph because Shape/Concat/Transpose
    # must quantize inputs with some default quantization options

    # In QAT testing in FHE is fast since there are no TLUs
    # For PTQ we only test that the model can be compiled and that it can be executed
```

**Review by RomanBredehoft — DISMISSED** [review]
some small details to be cleaned + I have a few questions, but overall looks good !

File: src/concrete/ml/onnx/ops_impl.py (hunk 1 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:2068]: Indeed. But this is true for many of our numpy ops actually (i copied pasted from unsqueeze).

I will add a commit to fix the typings maybe.

File: src/concrete/ml/quantization/quantized_ops.py (hunk 4 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_ops.py:2516]: Hmm not sure. Again a copy paste here, this assert comes back 5 times in this file.

File: tests/quantization/test_quantized_ops.py (hunk 5 above)
Comment (jfrery) [reply tests/quantization/test_quantized_ops.py:1153]: Good question. I don't think we can cover all edge cases. But when we see one we can add it to the list.

File: tests/quantization/test_quantized_ops.py (hunk 6 above)
Comment (jfrery) [reply tests/quantization/test_quantized_ops.py:1255]: Yes right I should have removed that. I used a fix seed for debugging. I should probably use pytest args instead. I will remove.

File: tests/torch/test_compile_torch.py (hunk 7 above)
Comment (jfrery) [reply tests/torch/test_compile_torch.py:1150]: Yes! Let's keep it consistent. I will update.

File: tests/torch/test_compile_torch.py (hunk 8 above)
Comment (jfrery) [reply tests/torch/test_compile_torch.py:1158]: Well with 3 bits we can't expect the values to be the same. However, we are only soing shape transformation so I guess that only the output shape should be checked. Not sure why we check the actual prediction at all.

File: src/concrete/ml/onnx/ops_impl.py (hunk 3 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:2081]: I can change the name yes

File: src/concrete/ml/onnx/ops_impl.py (hunk 1 above)
Comment (RomanBredehoft) [reply src/concrete/ml/onnx/ops_impl.py:2068]: yeah not surprised that some old stuff are not up to date with our typing / doc conventions !

File: src/concrete/ml/quantization/quantized_ops.py (hunk 4 above)
Comment (RomanBredehoft) [reply src/concrete/ml/quantization/quantized_ops.py:2516]: sure yes, it just seems a bit fishy, I think having a == should e fine (and better) !

File: tests/torch/test_compile_torch.py (hunk 8 above)
Comment (RomanBredehoft) [reply tests/torch/test_compile_torch.py:1158]: I see, well could you maybe improve that ? because I believe that when `is_qat` is False (which is your case with your added model), we do not check the shape !

File: src/concrete/ml/onnx/ops_impl.py (hunk 1 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:2068]: Done

File: src/concrete/ml/onnx/ops_impl.py (hunk 2 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:2075]: Done

File: src/concrete/ml/onnx/ops_impl.py (hunk 3 above)
Comment (jfrery) [reply src/concrete/ml/onnx/ops_impl.py:2081]: Done

File: tests/quantization/test_quantized_ops.py (hunk 6 above)
Comment (jfrery) [reply tests/quantization/test_quantized_ops.py:1255]: Done

File: tests/torch/test_compile_torch.py (hunk 7 above)
Comment (jfrery) [reply tests/torch/test_compile_torch.py:1150]: Done
